keyring/backends/libsecret.py

```diff
@@ -0,0 +1,139 @@
+    from gi.repository import GLib
+    gi.require_version('Secret', '1')
+    from gi.repository import Secret
+    available = True
+except (AttributeError, ImportError, ValueError):
+    pass
+
+log = logging.getLogger(__name__)
+
+
+class Keyring(KeyringBackend):
+    """libsecret Keyring"""
+
+    appid = 'Python keyring library'
+    if available:
+        schema = Secret.Schema.new(
+            "org.freedesktop.Secret.Generic",
+            Secret.SchemaFlags.NONE,
+            {
+                "application": Secret.SchemaAttributeType.STRING,
+                "service": Secret.SchemaAttributeType.STRING,
+                "username": Secret.SchemaAttributeType.STRING,
+            }
+        )
+
+    @properties.ClassProperty
+    @classmethod
+    def priority(cls):
+        with ExceptionRaisedContext() as exc:
+            Secret.__name__
+        if exc:
+            raise RuntimeError("libsecret required")
+        return 5
+
+    def get_password(self, service, username):
+        """Get password of the username for the service"""
+        attributes = {
+            "application": self.appid,
+            "service": service,
+            "username": username,
+        }
+        try:
+            items = Secret.password_search_sync(self.schema, attributes,
+                                                Secret.SearchFlags.UNLOCK, None)
+        except GLib.Error as error:
+            quark = GLib.quark_try_string('g-io-error-quark')
+            if error.matches(quark, Gio.IOErrorEnum.FAILED):
+                raise KeyringLocked('Failed to unlock the item!')
+            raise error
+        for item in items:
+            try:
+                return item.retrieve_secret_sync().get_text()
+            except GLib.Error as error:
+                quark = GLib.quark_try_string('secret-error')
+                if error.matches(quark, Secret.Error.IS_LOCKED):
+                    raise KeyringLocked('Failed to unlock the item!')
+                raise error
+
+    def set_password(self, service, username, password):
+        """Set password for the username of the service"""
+        collection = Secret.COLLECTION_DEFAULT
+        attributes = {
+            "application": self.appid,
+            "service": service,
+            "username": username,
+        }
+        label = "Password for '{}' on '{}'".format(username, service)
+        try:
+            stored = Secret.password_store_sync(self.schema, attributes, collection,
+                                                label, password, None)
+        except GLib.Error as error:
+            quark = GLib.quark_try_string('secret-error')
+            if error.matches(quark, Secret.Error.IS_LOCKED):
+                raise KeyringLocked("Failed to unlock the collection!")
+            quark = GLib.quark_try_string('g-io-error-quark')
+            if error.matches(quark, Gio.IOErrorEnum.FAILED):
+                raise KeyringLocked("Failed to unlock the collection!")
```

Comment:
Does `Gio.IOErrorEnum.FAILED` really mean unlock error? How does it differ from `Secret.Error.IS_LOCKED`?

Reply:
The IO error is raised when a user cancels flatpak's keyring-unlock request.